src/vahti.rs

```diff
@@ -28,159 +56,136 @@ pub struct VahtiItem {
-        return crate::tori::api::is_valid_url(url).await;
-    } else if HUUTONET_REGEX.is_match(url) {
-        return crate::huutonet::api::is_valid_url(url).await;
+impl Torimies {
+    pub async fn update_all_vahtis(&mut self) -> Result<(), Error> {
+        let vahtis = self.database.fetch_all_vahtis().await?;
+        self.update_vahtis(vahtis).await?;
+        Ok(())
     }
-    false
-}
 
-pub async fn update_all_vahtis(
-    db: Database,
-    itemhistory: Arc<Mutex<ItemHistory>>,
-    http: Arc<Http>,
-) -> Result<(), anyhow::Error> {
-    itemhistory.lock().await.purge_old();
-    let vahtis = db.fetch_all_vahtis_group().await?;
-    update_vahtis(db, itemhistory, http, vahtis).await?;
-    Ok(())
-}
+    pub async fn update_vahtis(&mut self, vahtis: Vec<DbVahti>) -> Result<(), Error> {
+        info!("Updating {} vahtis", vahtis.len());
+        let start = std::time::Instant::now();
 
-pub async fn update_vahtis(
-    db: Database,
-    itemhistory: Arc<Mutex<ItemHistory>>,
-    httpt: Arc<Http>,
-    grouped_vahtis: BTreeMap<String, Vec<Vahti>>,
-) -> Result<(), anyhow::Error> {
-    for (_, vahtis) in grouped_vahtis {
-        let http = httpt.clone();
-        let db = db.clone();
-        let itemhistory = itemhistory.clone();
-        tokio::spawn(async move {
-            let res = reqwest::get(vahtis[0].to_api().unwrap() + "&lim=10")
-                .await
-                .unwrap()
-                .text()
-                .await
-                .unwrap();
-            let site_id = SiteId::from(vahtis[0].url.as_str());
-            for vahti in vahtis {
-                if let Ok(mut currentitems) = vahti.parse_after_from_text(&res).await {
-                    if currentitems.len() == 10 {
-                        debug!("Unsure on whether we got all the items... Querying for all of them now");
-                        currentitems = vahti.parse_after().await.unwrap();
-                    }
-                    if currentitems.is_empty() {
-                        continue;
+        let ihs = self.itemhistorystorage.clone();
+
+        let db = self.database.clone();
+        let dm = self.delivery.clone();
+
+        // FIXME: Not a very good solution
+        for vahti in &vahtis {
+            if vahti.site_id == crate::tori::ID && ihs.get(&vahti.id).is_none() {
+                let ih = Arc::new(Mutex::new(ItemHistory::new()));
+
+                ihs.insert(vahti.id, ih);
+            }
+        }
+
+        let items = stream::iter(vahtis.iter().cloned())
+            .map(|v| (v, ihs.clone(), db.clone()))
+            .map(async move |(v, ihs, db)| match v.site_id {
+                #[cfg(feature = "tori")]
+                crate::tori::ID => {
+                    let vid = v.id;
+                    let Ok(mut tv) = ToriVahti::from_db(v) else {
+                            return vec![];
+                        };
+                    tv.itemhistory = ihs.get(&vid).map(|ih| ih.clone());
```

Comment:
.cloned()

Reply:
the output of the get function is of type `Option<Ref<i32, Arc<Mutex<ItemHistory>>>>` which is not an iterator and therefore doesn't implement .cloned()

Reply:
Right it was an option, not an iterator, this is all good